libbeat/common/wait/timer.go

```diff
@@ -0,0 +1,124 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package wait
+
+import (
+	"math/rand"
+	"time"
+)
+
+// Timer represents a timer implementation.
+type Timer interface {
+	Reset(time.Duration)
+	Start()
+	Stop()
+	Wait() <-chan time.Time
+}
+
+// Waiter is the strategy to be used to find out the time to wait before making a call.
+type Waiter func() time.Duration
+
+// MinWaitAndJitter takes a minimal wait time and jitter and will return a duration that will be at
+// least equal of bigger than the initial wait time.
+func MinWaitAndJitter(d, jitter time.Duration) Waiter {
+	return func() time.Duration {
+		return d + time.Duration(rand.Int63n(int64(jitter)))
+	}
+}
```

Comment:
Would you consider exponential backoff to be implemented on top of this package or as a strategy? In the laster case a strategy can not be a function.
Maybe a backoff implementation would just reuse the strategy, but not the timer?

Reply:
I think I would compose on top of it instead of defining a new strategy, backoff are direclty linked to the nature of calls:

- Retry?
- Abort?
- Try something else?

Buy building on top of it it allows to have the package with different concerns.